### PR TITLE
Compute area-weighted statistics

### PR DESCRIPTION
**Pull request by vincentsarago**
ref https://github.com/NASA-IMPACT/veda-backend/issues/226

This PR adds a `coverage` options to the `get_array_statistics` function to enable `area weighted` statistics. This PR do not take care of the `coverage` array creation which should be done by the client application (maybe with some helper in rio-tiler)

cc @kylebarron @j08lue

### To Do

- [x] more tests
- [x] docs
- [x] validate the logic (why some values are defined using weight or not)
- [x] add `coverage` utility functions

### CODE REVIEW

File: rio_tiler/utils.py [hunk 1]
```diff
@@ -36,6 +36,7 @@ def get_array_statistics(
     categorical: bool = False,
     categories: Optional[List[float]] = None,
     percentiles: Optional[List[int]] = None,
+    coverage: Optional[numpy.ndarray] = None,
```
Comment (kylebarron) [inline rio_tiler/utils.py:39]: Since I like really like typing 🙈 maybe
```py
from numpy.typing import NDArray
...
coverage: Optional[NDArray[np.floating]] = None
```
I think from reading this, it's expected to be a float? Even if it doesn't impact type checking much, I always find it helpful to recall what data types of arrays are expected where

File: rio_tiler/utils.py [hunk 2]
```diff
@@ -121,21 +122,36 @@ def get_array_statistics(
...
-                "sum": float(data[b].sum()),
-                "std": float(data[b].std()),
-                "median": float(numpy.ma.median(data[b])),
-                "majority": float(keys[counts.tolist().index(counts.max())].tolist())
-                if valid_pixels
-                else numpy.nan,
-                "minority": float(keys[counts.tolist().index(counts.min())].tolist())
-                if valid_pixels
-                else numpy.nan,
+                "mean": float(array.mean()),
```
Comment (vincentsarago) [inline rio_tiler/utils.py:148]: we don't use the weighted array for the `min` and `max` (like https://github.com/isciences/exactextract#supported-statistics)

File: rio_tiler/utils.py [hunk 3]
```diff
@@ -121,21 +122,36 @@ def get_array_statistics(
...
+        else:
+            array = data[b]
+
+        count = array.count() if coverage is None else coverage.sum()
+        if valid_pixels:
+            majority = float(keys[counts.tolist().index(counts.max())].tolist())
+            minority = float(keys[counts.tolist().index(counts.min())].tolist())
+        else:
+            majority = numpy.nan
+            minority = numpy.nan
```
Comment (vincentsarago) [inline rio_tiler/utils.py:142]: This is unrelated changes for styling

File: rio_tiler/utils.py [hunk 4]
```diff
@@ -121,21 +122,36 @@ def get_array_statistics(
...
-                else numpy.nan,
+                "mean": float(array.mean()),
+                "count": float(count),
+                "sum": float(array.sum()),
+                "std": float(array.std()),
+                "median": float(numpy.ma.median(array)),
+                "majority": majority,
+                "minority": minority,
                 "unique": float(counts.size),
                 **dict(zip(percentiles_names, percentiles_values)),
```
Comment (vincentsarago) [inline rio_tiler/utils.py:156]: 🤔 maybe the percentiles should use the weighted array?

File: tests/test_utils.py [hunk 5]
```diff
@@ -561,3 +561,30 @@ def parse(content):
...
+
+
+def test_get_array_statistics_coverage():
+    """Test statistics with coverage array."""
+    # Data Array
+    # 1, 2
+    # 3, 4
+    data = np.ma.array((1, 2, 3, 4)).reshape((1, 2, 2))
+
+    # Coverage Array
```
Comment (kylebarron) [inline tests/test_utils.py:573]: Maybe we should call this the `weight` array instead of `coverage`? `coverage` doesn't seem to make the most sense to me here.

**Review by kylebarron — APPROVED** [review]

File: tests/test_utils.py (hunk 5 above)
Comment (vincentsarago) [reply tests/test_utils.py:573]: in https://github.com/isciences/exactextract, weight are really weight, while the `coverage` is called `cell coverage fractions`. I didn't want to confuse people 🤷

Comment (kylebarron) [reply tests/test_utils.py:573]: I see, `coverage` came from exactextract because their weights are usually derived from the coverage of each polygon in the cell. I think `weight` is more general than `coverage` though. There could be use cases for weighted zonal stats that aren't partial-coverage

Comment (vincentsarago) [reply tests/test_utils.py:573]: `coverage_weights` ?

Comment (kylebarron) [reply tests/test_utils.py:573]: Would we ever have both `coverage` and `weights`?

Comment (kylebarron) [reply tests/test_utils.py:573]: It looks like exactextract does support both `coverage` and `weight` in effect, because it allows `weight` as a parameter and it computes `coverage` itself

Comment (vincentsarago) [reply tests/test_utils.py:573]: > Would we ever have both coverage and weights?

Not planned right now but I don't want to be blocked in the future. I agree that `coverage` is not a perfect name but I don't want to use `weights` because it's not it's not `weights` but spatial fraction. I'm open to change to better name if you have ideas :D

File: rio_tiler/models.py [hunk 6]
```diff
@@ -787,3 +790,43 @@ def statistics(
...
+            transform=self.transform * Affine.scale(1 / cover_scale),
+            all_touched=True,
+            fill=0,
+            dtype="uint8",
+        )
+        cover_array = cover_array.reshape(
+            (self.height, cover_scale, self.width, cover_scale)
+        ).astype("float32")
+
+        return cover_array.sum(-1).sum(1) / (cover_scale**2)
```
Comment (vincentsarago) [inline rio_tiler/models.py:832]: slightly modified version of https://github.com/perrygeo/python-rasterstats/pull/136

@sgoodm I know ☝️ is a 7 years old PR (😅) but I hope you don't mind that I reused some of the code here 🙏

**Comment by vincentsarago** [comment]
Note: we don't have a BaseReader method that returns statistics for a GeoJSON Feature but this is how it could look like in the `client` (e.g TiTiler):

```python
with Reader(path) as src:
    data = src_dst.feature(
        shape,
        shape_crs=WGS84_CRS,
    )

    coverage_array = data.get_coverage_array(
        shape, shape_crs=WGS84_CRS
    )

    stats = data.statistics(coverage=coverage_array)

```

File: rio_tiler/models.py (hunk 6 above)
Comment (sgoodm) [reply rio_tiler/models.py:833]: Not at all, @jacobwhall and I have been pushing use of COGs in our work at @aiddata and are fans of these projects, so happy to share some code to support :+1: :+1:

**Review by kylebarron — APPROVED** [review]
